File: frontend/Features/Community/KaleidoscopeDrawView.swift
import SwiftUI

// MARK: - 万花筒抽取浮层（原型 .kal）
// 选「类似 / 相反」→ 万花筒旋转 → 转出一位真实旅人 → 查看主页。

struct KaleidoscopeDrawView: View {
    @Environment(SupabaseService.self) private var supabase
    @Environment(\.dismiss) private var dismiss
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    enum Phase { case choose, spinning, result }
    enum Mode: String { case similar, opposite }

    @State private var phase: Phase = .choose
    @State private var mode: Mode = .similar
    @State private var spinTrigger = 0
    @State private var drawn: Traveler?
    /// 后端 AI 抽取给出的匹配理由（本地兜底时为 nil）
    @State private var drawReason: String?

    private var pool: [Traveler] {
        supabase.travelers.isEmpty ? DemoData.travelers : supabase.travelers
    }

    var body: some View {
        ZStack {
            background.ignoresSafeArea()
            VStack(spacing: 0) {
                header.padding(.top, 90)
                Spacer(minLength: 20)
                content
                Spacer(minLength: 20)
                actions.padding(.bottom, 50)
            }
            .padding(.horizontal, 28)
            closeButton
        }
    }

    private var background: some View {
        ZStack {
            Color(hex: 0x06080E, alpha: 0.97)
            RadialGradient(colors: [Color(hex: 0x5E96FF, alpha: 0.075), .clear], center: UnitPoint(x: 0.5, y: 0.30), startRadius: 0, endRadius: 300)
            RadialGradient(colors: [Color(hex: 0xE35CC1, alpha: 0.035), .clear], center: UnitPoint(x: 0.5, y: 0.90), startRadius: 0, endRadius: 240)
        }
        .background(.ultraThinMaterial)
    }

    private var closeButton: some View {
        VStack {
            HStack {
                Spacer()
                Button("关闭 ✕") { dismiss() }
                    .font(.system(size: 12.5)).foregroundStyle(Theme.faint)
                    .padding(.top, 62).padding(.trailing, 20)
            }
            Spacer()
        }
    }

    // MARK: 文案

    private var header: some View {
        VStack(spacing: 8) {
            Text(title).font(.system(size: 20, weight: .bold)).tracking(0.8).foregroundStyle(Theme.ink)
                .multilineTextAlignment(.center)
            if let sub {
                Text(sub).font(.system(size: 12)).foregroundStyle(Theme.sub).lineSpacing(4)
                    .multilineTextAlignment(.center)
            }
        }
    }

    private var title: String {
        switch phase {
        case .choose: return "你想看见哪一面的人生？"
        case .spinning: return mode == .similar ? "寻找和你同路的人" : "寻找你没走过的路"
        case .result: return mode == .similar ? "与你的画像最相似的人" : "与你的轨迹完全相反的人"
        }
    }

    private var sub: String? {
        switch phase {
        case .choose: return "万花筒会为你转出一位真实的旅人"
        case .spinning: return nil
        case .result: return drawReason ?? "看看 TA 的人生轨迹"
        }
    }

    // MARK: 主内容

    @ViewBuilder
    private var content: some View {
        switch phase {
        case .choose: modeOptions
        case .spinning:
            VStack(spacing: 20) {
                KaleidoscopeView(spinTrigger: spinTrigger)
                Text("光在旋转，人生在折叠……").font(.system(size: 12)).tracking(1.5).foregroundStyle(Theme.faint)
            }
        case .result:
            if let drawn { resultCard(drawn) }
        }
    }

    private var modeOptions: some View {
        HStack(spacing: 12) {
            modeCard(.similar, emoji: "🪞", title: "和我有类似经历", desc: "在同路人身上\n看见自己的下一步")
            modeCard(.opposite, emoji: "🔭", title: "和我经历完全相反", desc: "在另一种人生里\n看见没走过的路")
        }
    }

    private func modeCard(_ m: Mode, emoji: String, title: String, desc: String) -> some View {
        let on = mode == m
        return Button {
            withAnimation(.easeOut(duration: 0.2)) { mode = m }
        } label: {
            VStack(spacing: 0) {
                Text(emoji).font(.system(size: 22))
                Text(title).font(.system(size: 13.5, weight: .semibold)).foregroundStyle(Theme.ink).padding(.top, 10)
                Text(desc).font(.system(size: 11)).foregroundStyle(Theme.sub).lineSpacing(3)
                    .multilineTextAlignment(.center).padding(.top, 5)
            }
            .frame(maxWidth: .infinity)
            .padding(.horizontal, 14).padding(.vertical, 20)
            .background(
                on
                    ? AnyShapeStyle(LinearGradient(colors: [Color(hex: 0x5E96FF, alpha: 0.16), Color(hex: 0xE35CC1, alpha: 0.1)], startPoint: .topLeading, endPoint: .bottomTrailing))
                    : AnyShapeStyle(Theme.card),
                in: RoundedRectangle(cornerRadius: 20, style: .continuous)
            )
            .overlay(RoundedRectangle(cornerRadius: 20, style: .continuous).strokeBorder(on ? Color(hex: 0x6FA5FF, alpha: 0.7) : Theme.line, lineWidth: 1.5))
            .offset(y: on ? -2 : 0)
        }
        .buttonStyle(PressScaleStyle())
    }

    private func resultCard(_ t: Traveler) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            HueBandHeader(initial: t.initial, hue: t.hue, bandHeight: 74, avatarSize: 52,
                          imageName: MockAvatar.name(for: t.id))
            VStack(alignment: .leading, spacing: 8) {
                Text(t.name).font(.system(size: 16, weight: .semibold)).foregroundStyle(Theme.ink)
                Text("「\(t.quote)」").font(.system(size: 13)).foregroundStyle(Theme.sub).lineSpacing(5)
                    .frame(maxWidth: .infinity, alignment: .leading)
                FlowLayout(spacing: 5) {
                    ForEach(t.tags, id: \.self) { TagPill(text: $0) }
                }
                .padding(.top, 4)
            }
            .padding(.horizontal, 20).padding(.top, 28).padding(.bottom, 20)
        }
        .background(Theme.card, in: RoundedRectangle(cornerRadius: 24, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 24, style: .continuous).strokeBorder(Theme.line, lineWidth: 1))
        .shadow(color: .black.opacity(0.5), radius: 30, y: 12)
        .transition(.scale(scale: 0.92).combined(with: .opacity))
    }

    // MARK: 底部动作

    @ViewBuilder
    private var actions: some View {
        switch phase {
        case .choose:
            PrimaryButton(title: "开始转动 · 抽一位旅人") { spin() }
        case .spinning:
            EmptyView()
        case .result:
            VStack(spacing: 10) {
                if let drawn {
                    TravelerProfileLink(travelerId: drawn.id) {
                        Text("查看 TA 的主页")
                            .font(.system(size: 15, weight: .semibold)).foregroundStyle(.white)
                            .frame(maxWidth: .infinity).padding(.vertical, 15)
                            .background(Theme.buttonGradient, in: Capsule())
                            .shadow(color: Color(hex: 0x4F7DFF, alpha: 0.6), radius: 14, y: 8)
                    }
                }
                Button("再转一次") { withAnimation { phase = .choose; drawn = nil; drawReason = nil } }
                    .font(.system(size: 12.5)).foregroundStyle(Theme.faint).padding(.top, 2)
            }
        }
    }

    // MARK: 抽取

    private func spin() {
        withAnimation(.easeInOut(duration: 0.3)) { phase = .spinning }
        spinTrigger += 1
        let delay: Duration = reduceMotion ? .milliseconds(400) : .milliseconds(1450)
        let modeParam = mode == .similar ? "similar" : "different"
        Task {
            // 动画和远程抽取并行，接口最多占用一个动画周期；
            // 超时立即回落本地候选，避免网络延迟把转盘停在中间态。
            async let spinWait: Void? = try? await Task.sleep(for: delay)
            let picked = await withTaskGroup(
                of: (travelerId: Int, reason: String)?.self
            ) { group in
                group.addTask {
                    try? await supabase.kaleidoscopeDraw(mode: modeParam)
                }
                group.addTask {
                    try? await Task.sleep(for: delay)
                    return nil
                }
                let first = await group.next() ?? nil
                group.cancelAll()
                return first
            }
            _ = await spinWait

            let chosen: Traveler?
            if let picked, let t = pool.first(where: { $0.id == picked.travelerId }) {
                chosen = t
                drawReason = picked.reason
            } else {
                let candidates = pool.filter { mode == .similar ? $0.isSimilar : !$0.isSimilar }
                chosen = (candidates.isEmpty ? pool : candidates).randomElement()
                drawReason = nil
            }
            withAnimation(.spring(response: 0.5, dampingFraction: 0.75)) {
                drawn = chosen
                phase = .result
            }
        }
    }
}

#Preview {
    KaleidoscopeDrawView()
        .environment(SupabaseService())
        .preferredColorScheme(.dark)
}

File: android/app/src/main/java/app/possibility/android/features/community/KaleidoscopeDrawView.kt
package app.possibility.android.features.community

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import app.possibility.android.core.model.DemoData
import app.possibility.android.core.model.Traveler
import app.possibility.android.core.network.SupabaseService
import app.possibility.android.core.theme.Theme
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

// MARK: - 万花筒抽取浮层（原型 .kal）
// 选「类似 / 相反」→ 万花筒旋转 → 转出一位真实旅人 → 查看主页。
// 对应 ios/Possibility/Features/Community/KaleidoscopeDrawView.swift。

private enum class DrawPhase { CHOOSE, SPINNING, RESULT }
private enum class DrawMode(val api: String) { SIMILAR("similar"), OPPOSITE("different") }

/**
 * 万花筒抽取全屏浮层。
 * @param onPick 抽到旅人后点「看 TA」的回调（由外层就地打开旅人主页）。
 * @param onDismiss 关闭浮层。
 */
@Composable
fun KaleidoscopeDrawSheet(onPick: (Traveler) -> Unit, onDismiss: () -> Unit) {
    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false),
    ) {
        val service = SupabaseService.shared
        val scope = rememberCoroutineScope()

        var phase by remember { mutableStateOf(DrawPhase.CHOOSE) }
        var mode by remember { mutableStateOf(DrawMode.SIMILAR) }
        var drawn by remember { mutableStateOf<Traveler?>(null) }
        var reason by remember { mutableStateOf<String?>(null) }

        fun pool(): List<Traveler> =
            service.travelers.value.ifEmpty { DemoData.travelers }

        fun spin() {
            phase = DrawPhase.SPINNING
            scope.launch {
                val picked = runCatching { service.kaleidoscopeDraw(mode.api) }.getOrNull()
                delay(1450)
                val chosen: Traveler?
                val hit = picked?.let { p -> pool().firstOrNull { it.id == p.travelerId } }
                if (hit != null) {
                    chosen = hit
                    reason = picked.reason
                } else {
                    val candidates = pool().filter { if (mode == DrawMode.SIMILAR) it.isSimilar else !it.isSimilar }
                    chosen = (candidates.ifEmpty { pool() }).randomOrNull()
                    reason = null
                }
                drawn = chosen
                phase = DrawPhase.RESULT
            }
        }

        val title = when (phase) {
            DrawPhase.CHOOSE -> "你想看见哪一面的人生？"
            DrawPhase.SPINNING -> if (mode == DrawMode.SIMILAR) "寻找和你同路的人" else "寻找你没走过的路"
            DrawPhase.RESULT -> if (mode == DrawMode.SIMILAR) "与你的画像最相似的人" else "与你的轨迹完全相反的人"
        }
        val sub = when (phase) {
            DrawPhase.CHOOSE -> "万花筒会为你转出一位真实的旅人"
            DrawPhase.SPINNING -> null
            DrawPhase.RESULT -> reason ?: "看看 TA 的人生轨迹"
        }

        Box(
            Modifier
                .fillMaxSize()
                .background(Color(0xFF06080E).copy(alpha = 0.98f)),
        ) {
            // 背景弱光
            Box(
                Modifier
                    .fillMaxSize()
                    .background(Brush.radialGradient(listOf(Theme.blue.copy(alpha = 0.075f), Color.Transparent))),
            )

            // 关闭
            Row(
                Modifier
                    .fillMaxWidth()
                    .statusBarsPadding()
                    .padding(top = 20.dp, end = 20.dp),
                horizontalArrangement = Arrangement.End,
            ) {
                Text(
                    "关闭 ✕",
                    color = Theme.faint,
                    fontSize = 12.5.sp,
                    modifier = Modifier.communityClickable(onDismiss),
                )
            }

            Column(
                Modifier
                    .fillMaxSize()
                    .statusBarsPadding()
                    .padding(horizontal = 28.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Spacer(Modifier.height(90.dp))
                Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    Text(
                        title,
                        color = Theme.ink,
                        fontSize = 20.sp,
                        fontWeight = FontWeight.Bold,
                        letterSpacing = 0.8.sp,
                        textAlign = TextAlign.Center,
                    )
                    if (sub != null) {
                        Text(sub, color = Theme.sub, fontSize = 12.sp, lineHeight = 18.sp, textAlign = TextAlign.Center)
                    }
                }

                Spacer(Modifier.weight(1f))

                when (phase) {
                    DrawPhase.CHOOSE -> ModeOptions(mode) { mode = it }
                    DrawPhase.SPINNING -> SpinningOrb()
                    DrawPhase.RESULT -> drawn?.let { ResultCard(it) }
                }

                Spacer(Modifier.weight(1f))

                Box(Modifier.navigationBarsPadding().padding(bottom = 40.dp)) {
                    when (phase) {
                        DrawPhase.CHOOSE -> PrimaryPill("开始转动 · 抽一位旅人") { spin() }
                        DrawPhase.SPINNING -> Spacer(Modifier.height(1.dp))
                        DrawPhase.RESULT -> Column(
                            horizontalAlignment = Alignment.CenterHorizontally,
                            verticalArrangement = Arrangement.spacedBy(10.dp),
                        ) {
                            drawn?.let { t -> PrimaryPill("查看 TA 的主页") { onPick(t) } }
                            Text(
                                "再转一次",
                                color = Theme.faint,
                                fontSize = 12.5.sp,
                                modifier = Modifier.communityClickable {
                                    phase = DrawPhase.CHOOSE
                                    drawn = null
                                    reason = null
                                },
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun ModeOptions(mode: DrawMode, onSelect: (DrawMode) -> Unit) {
    Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
        ModeCard(
            selected = mode == DrawMode.SIMILAR,
            emoji = "🪞",
            title = "和我有类似经历",
            desc = "在同路人身上\n看见自己的下一步",
            modifier = Modifier.weight(1f),
        ) { onSelect(DrawMode.SIMILAR) }
        ModeCard(
            selected = mode == DrawMode.OPPOSITE,
            emoji = "🔭",
            title = "和我经历完全相反",
            desc = "在另一种人生里\n看见没走过的路",
            modifier = Modifier.weight(1f),
        ) { onSelect(DrawMode.OPPOSITE) }
    }
}

@Composable
private fun ModeCard(
    selected: Boolean,
    emoji: String,
    title: String,
    desc: String,
    modifier: Modifier = Modifier,
    onClick: () -> Unit,
) {
    Column(
        modifier
            .clip(RoundedCornerShape(20.dp))
            .background(
                if (selected) {
                    Brush.linearGradient(listOf(Theme.blue.copy(alpha = 0.16f), Theme.magenta.copy(alpha = 0.1f)))
                } else {
                    Brush.linearGradient(listOf(Theme.card, Theme.card))
                },
            )
            .border(
                1.5.dp,
                if (selected) Color(0xFF6FA5FF).copy(alpha = 0.7f) else Theme.line,
                RoundedCornerShape(20.dp),
            )
            .communityClickable(onClick)
            .padding(horizontal = 14.dp, vertical = 20.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(emoji, fontSize = 22.sp)
        Text(
            title,
            color = Theme.ink,
            fontSize = 13.5.sp,
            fontWeight = FontWeight.SemiBold,
            modifier = Modifier.padding(top = 10.dp),
            textAlign = TextAlign.Center,
        )
        Text(
            desc,
            color = Theme.sub,
            fontSize = 11.sp,
            lineHeight = 16.sp,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = 5.dp),
        )
    }
}

/** 旋转的万花筒球（原型 KaleidoscopeView 的简化版）。 */
@Composable
private fun SpinningOrb() {
    val transition = rememberInfiniteTransition(label = "spin")
    val angle by transition.animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(
            animation = androidx.compose.animation.core.tween(1400, easing = LinearEasing),
            repeatMode = RepeatMode.Restart,
        ),
        label = "spin-angle",
    )
    Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(20.dp)) {
        Box(
            Modifier
                .size(150.dp)
                .graphicsLayer { rotationZ = angle }
                .clip(CircleShape)
                .background(Brush.sweepGradient(Theme.orbConicColors))
                .border(2.dp, Color.White.copy(alpha = 0.2f), CircleShape),
        )
        Text("光在旋转，人生在折叠……", color = Theme.faint, fontSize = 12.sp, letterSpacing = 1.5.sp)
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun ResultCard(t: Traveler) {
    Column(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(24.dp))
            .background(Theme.card)
            .border(1.dp, Theme.line, RoundedCornerShape(24.dp)),
    ) {
        // 色带头
        Box(
            Modifier
                .fillMaxWidth()
                .height(74.dp)
                .background(Theme.hue(t.hue).gradient),
            contentAlignment = Alignment.Center,
        ) {
            Box(
                Modifier
                    .size(52.dp)
                    .clip(CircleShape)
                    .background(Theme.hue(t.hue).gradient)
                    .border(2.dp, Color.White.copy(alpha = 0.55f), CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Text(t.initial, color = Color.White, fontSize = 22.sp, fontWeight = FontWeight.Bold)
            }
        }
        Column(
            Modifier.padding(horizontal = 20.dp, vertical = 20.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Text(t.name, color = Theme.ink, fontSize = 16.sp, fontWeight = FontWeight.SemiBold)
            Text("「${t.quote}」", color = Theme.sub, fontSize = 13.sp, lineHeight = 20.sp)
            FlowRow(horizontalArrangement = Arrangement.spacedBy(5.dp), verticalArrangement = Arrangement.spacedBy(5.dp)) {
                t.tags.forEach { CommunityTagPill(it) }
            }
        }
    }
}

@Composable
private fun PrimaryPill(label: String, onClick: () -> Unit) {
    Box(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(999.dp))
            .background(Theme.buttonGradient)
            .communityClickable(onClick)
            .padding(vertical = 15.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(label, color = Color.White, fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
    }
}
